src/librustc_trait_selection/traits/wf.rs

```diff
@@ -504,8 +519,9 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
                     // are not directly inspecting closure types
                     // anyway, except via auto trait matching (which
                     // only inspects the upvar types).
-                    subtys.skip_current_subtree(); // subtree handled by compute_projection
+                    walker.skip_current_subtree(); // subtree handled below
                     for upvar_ty in substs.as_closure().upvar_tys() {
+                        // FIXME(eddyb) add the type to `walker` instead of recursing.
                         self.compute(upvar_ty);
```

Comment:
Wait I forgot about this, this is technically a bugfix (before, if `compute` returned `false` here, those inference variables would be forgotten about).

So it might make sense that it's slower. Let me, uh, make a PR that only changes this, and check that for perf regressions.

Reply:
This is it.

`typeck_tables_of` on `clap-rs` gets slightly faster (instead of being a significant regression), if I ignore upvar types that are `ty::Infer(ty::TyVar(_))` and don't resolve.

But that's a WF hole, right? Do we just accept the regression?

Reply:
Sorry, I lost track of this PR. I agree that this is a bug-fix.